File: src/config/part_compatibility.cpp
#include "part_compatibility.hpp"
#include <fstream>
#include <sstream>

static std::string Trim(const std::string &s) {
  size_t start = 0;
  while (start < s.size() && (s[start] == ' ' || s[start] == '\t'))
    start++;
  size_t end = s.size();
  while (end > start && (s[end - 1] == ' ' || s[end - 1] == '\t'))
    end--;
  return s.substr(start, end - start);
}

static void FinalizeRule(CompatibilityRule &rule,
                         std::vector<CompatibilityRule> &rules) {
  if (rule.ifSlot.empty() || rule.ifPart.empty())
    return;
  rules.push_back(rule);
}

std::vector<CompatibilityRule>
LoadPartCompatibility(const std::string &filename) {
  std::vector<CompatibilityRule> rules;
  std::ifstream file(filename);
  if (!file.is_open())
    return rules;

  CompatibilityRule current;
  std::string line;
  while (std::getline(file, line)) {
    if (line.empty() || line[0] == '#')
      continue;

    auto eq = line.find('=');
    if (eq == std::string::npos)
      continue;

    std::string key = Trim(line.substr(0, eq));
    std::string value = Trim(line.substr(eq + 1));

    if (key == "rule") {
      FinalizeRule(current, rules);
      current = CompatibilityRule{};
      continue;
    }

    if (key == "if_slot")
      current.ifSlot = value;
    else if (key == "if_part")
      current.ifPart = value;
    else if (key == "requires_slot") {
      current.kind = CompatibilityRule::Kind::Requires;
      current.otherSlot = value;
    } else if (key == "requires_part")
      current.otherPart = value;
    else if (key == "requires_any_parts") {
      current.kind = CompatibilityRule::Kind::RequiresAny;
      std::istringstream fields(value);
      std::string item;
      while (std::getline(fields, item, ',')) {
        item = Trim(item);
        if (!item.empty())
          current.otherPartsAny.push_back(item);
      }
    } else if (key == "forbids_slot") {
      current.kind = CompatibilityRule::Kind::Forbids;
      current.otherSlot = value;
    } else if (key == "forbids_part")
      current.otherPart = value;
  }

  FinalizeRule(current, rules);
  return rules;
}

bool ValidatePartCompatibility(const CarConfig &car,
                               const std::vector<CompatibilityRule> &rules,
                               std::string *errorOut) {
  if (car.fuelSystemId == "HydrogenTank" && car.engine.fuelType != "Hydrogen") {
    if (errorOut) {
      *errorOut =
          "fuel_system.HydrogenTank requires Hydrogen fuel in the powertrain";
    }
    return false;
  }

  if (car.engine.fuelType == "Hydrogen" &&
      car.engine.energyConverter == "FuelCell") {
    if (car.engine.drivetrain != "FullEV") {
      if (errorOut) {
        *errorOut =
            "Hydrogen fuel cell requires FullEV drivetrain (SingleSpeedEDrive)";
      }
      return false;
    }
    if (car.hybridSystemId != "None" && !car.hybridSystemId.empty()) {
      if (errorOut) {
        *errorOut = "Fuel cell powertrain cannot use a separate hybrid system";
      }
      return false;
    }
    if (car.transmissionId != "SingleSpeedEDrive") {
      if (errorOut) {
        *errorOut =
            "Hydrogen fuel cell requires SingleSpeedEDrive transmission";
      }
      return false;
    }
  }

  if (car.engine.fuelType == "Hydrogen" &&
      car.engine.drivetrain == "RangeExtender") {
    if (errorOut) {
      *errorOut =
          "Hydrogen range-extender is not supported; use Fuel cell instead";
    }
    return false;
  }

  for (const CompatibilityRule &rule : rules) {
    if (PartChoiceForSlot(car, rule.ifSlot) != rule.ifPart)
      continue;

    const std::string otherChoice = PartChoiceForSlot(car, rule.otherSlot);
    if (rule.kind == CompatibilityRule::Kind::Requires) {
      if (otherChoice != rule.otherPart) {
        if (errorOut) {
          *errorOut = rule.ifSlot + "." + rule.ifPart + " requires " +
                      rule.otherSlot + "." + rule.otherPart + " (got " +
                      otherChoice + ")";
        }
        return false;
      }
    } else if (rule.kind == CompatibilityRule::Kind::RequiresAny) {
      bool matched = false;
      for (const std::string &allowed : rule.otherPartsAny) {
        if (otherChoice == allowed) {
          matched = true;
          break;
        }
      }
      if (!matched) {
        if (errorOut) {
          *errorOut = rule.ifSlot + "." + rule.ifPart + " requires one of " +
                      rule.otherSlot + " carbon-class options (got " +
                      otherChoice + ")";
        }
        return false;
      }
    } else if (otherChoice == rule.otherPart) {
      if (errorOut) {
        *errorOut = rule.ifSlot + "." + rule.ifPart + " forbids " +
                    rule.otherSlot + "." + rule.otherPart;
      }
      return false;
    }
  }
  return true;
}

File: server/src/game/part_compatibility.ts
import * as fs from "fs";
import * as path from "path";
import type { CarBuildPayload } from "../ws_protocol";
import {
  EV_ONLY_OUTLET_PARTS,
  isElectricDriveOutletBuild,
  isEvLegalOutlet,
} from "./ev_outlet";

/** Config slot names in part_compatibility.txt → CarBuildPayload fields. */
export const BUILD_FIELD_BY_CONFIG_SLOT: Record<string, keyof CarBuildPayload> = {
  chassis: "chassis_type",
  front_aero: "front_aero_type",
  rear_aero: "rear_aero_type",
  diffuser: "diffuser_type",
  exhaust: "exhaust_type",
  cooling: "cooling_pack",
  wheel_package: "wheel_package",
  suspension: "suspension_layout",
  fuel_system: "fuel_system",
  brake_system: "brake_system",
  transmission: "transmission",
  hybrid_system: "hybrid_system",
};

export interface AssemblyRequiresAnyRule {
  kind: "requires_any";
  ifSlot: string;
  ifPart: string;
  requiresSlot: string;
  requiresAnyParts: string[];
}

export interface AssemblyRequiresRule {
  kind: "requires";
  ifSlot: string;
  ifPart: string;
  requiresSlot: string;
  requiresPart: string;
}

export type AssemblyRule = AssemblyRequiresAnyRule | AssemblyRequiresRule;

function trim(s: string): string {
  return s.trim();
}

export function loadAssemblyRules(repoRoot: string): AssemblyRule[] {
  const file = path.join(repoRoot, "configs/part_compatibility.txt");
  if (!fs.existsSync(file)) return [];

  const rules: AssemblyRule[] = [];
  type ParserState = {
    kind: "requires" | "requires_any";
    ifSlot?: string;
    ifPart?: string;
    requiresSlot?: string;
    requiresPart?: string;
    requiresAnyParts: string[];
  };
  let current: ParserState | null = null;

  const flush = () => {
    if (!current?.ifSlot || !current.ifPart || !current.requiresSlot) {
      current = null;
      return;
    }
    if (current.kind === "requires" && current.requiresPart) {
      rules.push({
        kind: "requires",
        ifSlot: current.ifSlot,
        ifPart: current.ifPart,
        requiresSlot: current.requiresSlot,
        requiresPart: current.requiresPart,
      });
    } else if (current.requiresAnyParts.length > 0) {
      rules.push({
        kind: "requires_any",
        ifSlot: current.ifSlot,
        ifPart: current.ifPart,
        requiresSlot: current.requiresSlot,
        requiresAnyParts: current.requiresAnyParts,
      });
    }
    current = null;
  };

  for (const line of fs.readFileSync(file, "utf8").split("\n")) {
    const trimmed = trim(line);
    if (!trimmed || trimmed.startsWith("#")) continue;
    const eq = trimmed.indexOf("=");
    if (eq < 0) continue;
    const key = trim(trimmed.slice(0, eq));
    const value = trim(trimmed.slice(eq + 1));

    if (key === "rule") {
      flush();
      current = { kind: "requires_any", requiresAnyParts: [] };
    } else if (!current) {
      continue;
    } else if (key === "if_slot") {
      current.ifSlot = value;
    } else if (key === "if_part") {
      current.ifPart = value;
    } else if (key === "requires_slot") {
      current.requiresSlot = value;
    } else if (key === "requires_part") {
      current.kind = "requires";
      current.requiresPart = value;
    } else if (key === "requires_any_parts") {
      current.kind = "requires_any";
      current.requiresAnyParts = value.split(",").map(trim).filter(Boolean);
    }
  }
  flush();
  return rules;
}

export function buildFieldValue(
  build: CarBuildPayload,
  configSlot: string,
): string {
  const field = BUILD_FIELD_BY_CONFIG_SLOT[configSlot];
  if (!field) return "";
  const value = build[field];
  if (typeof value === "string" && value) return value;
  if (field === "diffuser_type") return "StockFloor";
  if (field === "exhaust_type") return "TwinOutletSide";
  return "";
}

function validateFuelSystemPowertrain(build: CarBuildPayload): string | null {
  if (
    build.fuel_system === "HydrogenTank" &&
    build.engine?.fuel_type !== "Hydrogen"
  ) {
    return "Hydrogen tank requires Hydrogen fuel in the powertrain";
  }
  const eng = build.engine;
  if (eng?.fuel_type === "Hydrogen" && eng.energy_converter === "FuelCell") {
    if (build.hybrid_system && build.hybrid_system !== "None") {
      return "Fuel cell powertrain cannot use a separate hybrid system";
    }
    if (build.transmission && build.transmission !== "SingleSpeedEDrive") {
      return "Hydrogen fuel cell requires SingleSpeedEDrive transmission";
    }
    if (eng.drivetrain && eng.drivetrain !== "FullEV") {
      return "Hydrogen fuel cell requires FullEV drivetrain";
    }
  }
  if (eng?.fuel_type === "Hydrogen" && eng.drivetrain === "RangeExtender") {
    return "Hydrogen range-extender is not supported; use fuel cell instead";
  }
  return null;
}

const DPF_EXHAUST_PARTS = new Set(["DieselDPF", "DieselDPFSport"]);

function validateExhaustPowertrain(build: CarBuildPayload): string | null {
  const exhaust = build.exhaust_type ?? "TwinOutletSide";
  const eng = build.engine;
  if (DPF_EXHAUST_PARTS.has(exhaust) && eng?.fuel_type !== "Diesel") {
    return "Diesel DPF exhaust requires Diesel fuel in the powertrain";
  }
  const isEv = isElectricDriveOutletBuild(eng);
  if (isEv && !isEvLegalOutlet(exhaust)) {
    return "E-drive powertrain requires an underbody outlet package";
  }
  if (!isEv && (exhaust === "None" || EV_ONLY_OUTLET_PARTS.has(exhaust))) {
    return "Combustion powertrain requires an exhaust system";
  }
  return null;
}

export function validateAssemblyCompatibility(
  build: CarBuildPayload,
  rules: AssemblyRule[],
): string | null {
  for (const rule of rules) {
    if (buildFieldValue(build, rule.ifSlot) !== rule.ifPart) continue;

    const other = buildFieldValue(build, rule.requiresSlot);
    if (rule.kind === "requires") {
      if (other !== rule.requiresPart) {
        return `${rule.ifPart} requires ${rule.requiresPart} on ${rule.requiresSlot}`;
      }
    } else if (!rule.requiresAnyParts.includes(other)) {
      return `${rule.ifPart} is not compatible with ${other} on ${rule.requiresSlot}`;
    }
  }

  if (
    build.rear_aero_type === "WinglessGroundEffect" &&
    build.front_aero_type !== "LowDragNose" &&
    build.front_aero_type !== "LowDragNoseSlim"
  ) {
    return "Wingless rear package requires a low-drag nose";
  }

  if (
    build.rear_aero_type === "WinglessGroundEffect" &&
    (build.diffuser_type ?? "StockFloor") === "StockFloor"
  ) {
    return "Wingless rear requires a diffuser floor package";
  }

  const exhaustErr = validateExhaustPowertrain(build);
  if (exhaustErr) return exhaustErr;

  return validateFuelSystemPowertrain(build);
}

/** True when selecting `candidatePart` for `configSlot` keeps the build valid. */
export function isAssemblyPartCompatible(
  build: CarBuildPayload,
  configSlot: string,
  candidatePart: string,
  rules: AssemblyRule[],
): boolean {
  const field = BUILD_FIELD_BY_CONFIG_SLOT[configSlot];
  if (!field) return true;
  const preview = { ...build, [field]: candidatePart };
  return validateAssemblyCompatibility(preview, rules) === null;
}
